test/compilable/b16976.d

```diff
@@ -0,0 +1,23 @@
+/* REQUIRED_ARGS: -m64
+TEST_OUTPUT:
+---
+compilable/b16976.d(15): Deprecation: foreach: loop index implicitly converted from `size_t` to `int`
+compilable/b16976.d(16): Deprecation: foreach: loop index implicitly converted from `size_t` to `int`
+compilable/b16976.d(17): Deprecation: foreach: loop index implicitly converted from `size_t` to `char`
+compilable/b16976.d(18): Deprecation: foreach: loop index implicitly converted from `size_t` to `char`
+---
+*/
+void main()
+{
+    int[]  dyn = [1,2,3,4,5];
+    int[5] sta = [1,2,3,4,5];
+
+    foreach(int i, v; dyn) { }
+    foreach_reverse(int i, v; dyn) { }
+    foreach(char i, v; dyn) { }
+    foreach_reverse(char i, v; dyn) { }
+    foreach(int i, v; sta) { }
```

Comment:
I think we could/should factor in the size of the static array here.
E.g. in your case the maximum index of the array fits in an `int`, so why disallow it ?

Reply:
Yeah.

Reply:
If vrp can be done, all the better.  Continuing to allow any integral types seem reasonable to me if they can be statically checked as being safe.

Floating point and character types as indexes should be all errors (`foreach (dchar I, v; a)`).

Reply:
Floating point already fail (but with an ugly error message), `char` and friends don't.

Reply:
> I think we could/should factor in the size of the static array here.

Oh, turns out that that is handled in the code code literally right above my addition.